src/widgets/scrollbar.rs

```diff
@@ -498,25 +498,36 @@ impl Scrollbar<'_> {
     ///
     /// This method returns the length of the start, thumb, and end as a tuple.
     fn part_lengths(&self, area: Rect, state: &mut ScrollbarState) -> (usize, usize, usize) {
-        let track_len = self.track_length_excluding_arrow_heads(area) as f64;
-        let viewport_len = self.viewport_length(state, area) as f64;
+        const fn rounded_divide(a: usize, b: usize) -> usize {
+            (a + (b / 2)) / b
+        }
+
+        let track_len = self.track_length_excluding_arrow_heads(area) as usize;
+        let viewport_len = self.viewport_length(state, area) as usize;
 
-        let content_length = state.content_length as f64;
-        // Clamp the position to show at least one line of the content, even if the content is
-        let position = state.position.min(state.content_length - 1) as f64;
+        // Clamp the position to show at least one line of the content
+        // Note: state.content_length is != 0 because otherwise render() returns immediately
+        let position = state.position.min(state.content_length - 1);
 
         // vscode style scrolling behavior (allow scrolling past end of content)
-        let scrollable_content_len = content_length + viewport_len - 1.0;
-        let thumb_start = position * track_len / scrollable_content_len;
-        let thumb_end = (position + viewport_len) * track_len / scrollable_content_len;
+        let scrollable_content_len = state.content_length + viewport_len - 1;
+
+        // Calculate the thumb start (inclusive) and end (exclusive) positions, with rounding
+        let thumb_start = rounded_divide(position * track_len, scrollable_content_len);
+        let thumb_end = rounded_divide(
+            (position + viewport_len) * track_len,
+            scrollable_content_len,
+        );
 
-        // We round just the positions (instead of floor / ceil), and then calculate the sizes from
-        // those positions. Rounding the sizes instead causes subtle off by 1 errors.
-        let track_start_len = thumb_start.round() as usize;
-        let thumb_end = thumb_end.round() as usize;
+        // Ensure there is at least one thumb character by ensuring `thumb_start` has at least one
+        // space after it, and then ensuring `thumb_end` is at least one space after `thumb_start`
+        let thumb_start = thumb_start.min(track_len.saturating_sub(1));
+        let thumb_end = thumb_end.max(thumb_start + 1);
 
-        let thumb_len = thumb_end.saturating_sub(track_start_len);
-        let track_end_len = track_len as usize - track_start_len - thumb_len;
+        // Calculate the length of the tracks and thumb
+        let track_start_len = thumb_start;
+        let thumb_len = thumb_end - thumb_start;
+        let track_end_len = track_len - thumb_start - thumb_len;
 
         (track_start_len, thumb_len, track_end_len)
```

Comment:
I hear you on the floats - we've seen that a few times before - particularly in the layout, but I'm convinced that this code expresses the intent of the calculations much clearer than the int version as the rounding is explicit rather than. That is unless there's a specific case where the calculation breaks in ways that make this a problem. Here I think there's unlikely to be that (though this is probably the best way to be proven wrong by someone :)).